marconi-core/test/Marconi/Core/Spec/TracedSqlite.hs

```diff
@@ -0,0 +1,372 @@
+      let pts = map sePoint es
+      -- And run a fold over all of them, giving us all the historical values of the
+      -- indexer across all the stored values.
+      rs  <- fmap getResult <$> mapM (\pt -> Storable.query pt st (QEvents f)) pts
+      if null rs
+      then do
+        -- If there are no results, then return a single element list with the
+        -- accumulator.
+        Result ag0 <- Storable.query Genesis st QAccumulator
+        pure $ Just [ag0]
+      else
+        -- If there are results, return a list of all the values *without* the
+        -- accumulator.
+        pure . Just . take sz $ reverse rs
+
+getView
+  :: IndexT
+  -> PropertyM IO (Maybe (IndexView Int))
+getView ix = do
+  mix <- liftIO $ run nullTracer ix
+  case mix of
+    Nothing       -> pure Nothing
+    Just (ix', _) -> do
+      let maxSize = ix' ^. state . config . memoryBufferSize + 1
+      es <- liftIO $ Storable.getEvents (ix' ^. state)
+      Just rs <- getHistory ix
+      let sz = if null es then 1 else length es + 1
+      -- let sz = if length rs == 1 then 1 else length rs + 1
+      pure . Just $
+        IndexView { ixDepth = maxSize
+                  , ixView  = head rs
+                  , ixSize  = min maxSize sz
+                  }
+
+monadic
+  :: PropertyM IO Property
+  -> Property
+monadic = monadicIO
+
+{-
+   This function is needed to lookup a point from the model into the new version of
+   indexers. This is needed as rewinds in the model take a number of blocks to rewind
+   back, while in the new indexers they take a point that we want to rewind to.
+
+   This is not trivial because it has to respect both the rules surrounding rewinds
+   in the original model and the rules surrounding rewinds in the adapter version.
+-}
+lookupPoint
+  :: Int
+  -> Config
+  -> MaybeT IO (StorablePoint Handle)
+lookupPoint n (Config st _) = MaybeT $ do
+  let depth = st ^. config . memoryBufferSize + 1
+  es' <- reverse <$> Storable.getEvents st
+  -- In our model we may get a lot more events than `depth` since we never actually
+  -- remove them.
+  es  <- take (depth - 1) . reverse <$> Storable.getEvents st
+  -- Rewind past all stored events in the model.
+  if length es == n
+  then do
+    if length es' > length es
+       -- On this branch we know that we have more than `depth` events stored in the db, so
+       -- we do a rewind up to the `depth` + 1 element. This is semantically identical to what
+       -- happens in the model.
+       then pure . Just $ sePoint $ es' !! n
+       -- If we don't have more events stored in the database, then we simply rollback to
+       -- Genesis.
+       else pure . Just $ Genesis
+  else if length es > n
+  -- If the point we want to rollback to is less then the size of the list of stored
+  -- events, then we want to rollback to the indexed event.
+  then pure $ atMay es' n <&> sePoint
+  -- Otherwise the point we need to rollback to was not found.
+  else pure Nothing
```

Comment:
A few implementation remarks: `compare (length es) n` with a case case would arguably be more readable than nested `if then else` here, we could also use `splitAt` instead of `take` to check that the reminder is not null and avoid an extra call to `length`.

Reply:
I am not seeing all the changes at once right now, but I try to rewrite the code later.